### PR TITLE
Align Rng use in tests

### PR DESCRIPTION
**Pull request by ljedrz**
This PR is an extension of https://github.com/AleoHQ/snarkVM/pull/999.

1. introduce a `TestRng` that implements both `Rng` and `CryptoRng` (the latter of which is only a marker trait, so it's fine for test purposes where we desire reproducibility rather than true entropy)
2. replace the uses of `test_rng` and `test_crypto_rng` with `TestRng::default`
3. replace the uses of `test_crypto_rng_fixed` with `TestRng::fixed`
4. reuse the rngs instead of recreating them in all the tests
5. remove all direct uses of `XorShiftRng`
6. replace the uses of `thread_rng` in tests and benchmarks with `TestRng::default`

The benefits are reproducibility in case any of the related tests fails and greater performance.

### CODE REVIEW

**Comment by ljedrz** [comment]
It seems that one of the `x / MODULUS` (I haven't checked, but a comment states that the test can break on it) cases was just hit; it's reproducible on `testnet3`:
```
---- random::tests::test_random stdout ----

Initializing 'TestRng' with seed '11753855360983655227'

thread 'random::tests::test_random' panicked at '"D"', console/types/string/src/random.rs:51:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: circuit/algorithms/src/bhp/hash_uncompressed.rs [hunk 1]
```diff
@@ -80,9 +80,11 @@ mod tests {
             let native = console::$bhp::<<Circuit as Environment>::Network>::setup(DOMAIN)?;
             let circuit = $bhp::<Circuit>::constant(native.clone());
 
+            let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/bhp/hash_uncompressed.rs]: The macro wrapping this call is called multiple times in test cases lower down, thus creating the rng multiple times per test.

File: circuit/algorithms/src/pedersen/commit.rs [hunk 2]
```diff
@@ -113,11 +113,13 @@ mod tests {
         let native = console::Pedersen::<<Circuit as Environment>::Network, NUM_BITS>::setup(MESSAGE);
         let circuit = Pedersen::<Circuit, NUM_BITS>::constant(native.clone());
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/commit.rs]: Ditto

File: circuit/algorithms/src/pedersen/commit.rs [hunk 3]
```diff
@@ -160,9 +162,11 @@ mod tests {
     ) {
         println!("Checking homomorphic addition on {} + {}", first, second);
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/commit.rs]: Ditto

File: circuit/algorithms/src/pedersen/commit_uncompressed.rs [hunk 4]
```diff
@@ -123,11 +123,13 @@ mod tests {
         let native = console::Pedersen::<<Circuit as Environment>::Network, NUM_BITS>::setup(MESSAGE);
         let circuit = Pedersen::<Circuit, NUM_BITS>::constant(native.clone());
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/commit_uncompressed.rs]: Ditto

File: circuit/algorithms/src/pedersen/commit_uncompressed.rs [hunk 5]
```diff
@@ -166,9 +168,11 @@ mod tests {
     ) {
         println!("Checking homomorphic addition on {} + {}", first, second);
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/commit_uncompressed.rs]: Ditto

File: circuit/algorithms/src/pedersen/hash.rs [hunk 6]
```diff
@@ -66,9 +66,11 @@ mod tests {
         let native = console::Pedersen::<<Circuit as Environment>::Network, NUM_BITS>::setup(MESSAGE);
         let circuit = Pedersen::<Circuit, NUM_BITS>::constant(native.clone());
 
+        let rng = &mut TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/hash.rs]: Ditto

File: circuit/algorithms/src/pedersen/hash_uncompressed.rs [hunk 7]
```diff
@@ -124,9 +124,11 @@ mod tests {
         let native = console::Pedersen::<<Circuit as Environment>::Network, NUM_BITS>::setup(MESSAGE);
         let circuit = Pedersen::<Circuit, NUM_BITS>::constant(native.clone());
 
+        let rng = &mut TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/pedersen/hash_uncompressed.rs]: Ditto

File: circuit/algorithms/src/poseidon/hash.rs [hunk 8]
```diff
@@ -50,10 +50,12 @@ mod tests {
         let native = console::Poseidon::<<Circuit as Environment>::Network, RATE>::setup(DOMAIN)?;
         let poseidon = Poseidon::<Circuit, RATE>::constant(native.clone());
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/poseidon/hash.rs]: Ditto

File: circuit/algorithms/src/poseidon/hash_many.rs [hunk 9]
```diff
@@ -220,10 +220,12 @@ mod tests {
         let native = console::Poseidon::<<Circuit as Environment>::Network, { RATE as usize }>::setup(DOMAIN)?;
         let poseidon = Poseidon::<Circuit, { RATE as usize }>::constant(native.clone());
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/poseidon/hash_many.rs]: And again.

File: circuit/algorithms/src/poseidon/hash_to_scalar.rs [hunk 10]
```diff
@@ -57,9 +57,11 @@ mod tests {
         let native = console::Poseidon::<<Circuit as Environment>::Network, RATE>::setup(DOMAIN)?;
         let poseidon = Poseidon::<Circuit, RATE>::constant(native.clone());
 
+        let rng = &mut TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/poseidon/hash_to_scalar.rs]: Same here, on the plus side moving the rng out will guard against this inefficiency in future.

File: circuit/algorithms/src/poseidon/mod.rs [hunk 11]
```diff
@@ -23,7 +23,7 @@ mod prf;
 #[cfg(all(test, console))]
 use snarkvm_circuit_types::environment::assert_scope;
 #[cfg(test)]
-use snarkvm_utilities::{test_rng, Uniform};
+use snarkvm_utilities::{TestRng, Uniform};
```
Comment (niklaslong) [inline circuit/algorithms/src/poseidon/mod.rs:26]: I think this is unused.

File: circuit/algorithms/src/poseidon/prf.rs [hunk 12]
```diff
@@ -58,13 +58,15 @@ mod tests {
         let native = console::Poseidon::<<Circuit as Environment>::Network, RATE>::setup(DOMAIN)?;
         let poseidon = Poseidon::<Circuit, RATE>::constant(native.clone());
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/algorithms/src/poseidon/prf.rs]: Same case as earlier, moving out the rng creation from the function call would speed test cases up.

File: circuit/algorithms/src/poseidon/mod.rs (hunk 11 above)
Comment (ljedrz) [reply circuit/algorithms/src/poseidon/mod.rs:26]: IIRC it is used; the other modules rely on this import

File: circuit/program/src/data/record/decrypt.rs [hunk 13]
```diff
@@ -105,8 +105,10 @@ mod tests {
         owner: Owner<A, Plaintext<A>>,
         gates: Balance<A, Plaintext<A>>,
     ) -> Result<()> {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/program/src/data/record/decrypt.rs]: Same here, this function is called multiple times per test case.

File: circuit/types/address/src/equal.rs [hunk 14]
```diff
@@ -45,7 +45,7 @@ mod tests {
         num_private: u64,
         num_constraints: u64,
     ) {
-        let rng = &mut test_rng();
+        let rng = &mut TestRng::default();
```
Comment (niklaslong) [inline circuit/types/address/src/equal.rs]: Same here.

File: circuit/types/address/src/equal.rs [hunk 15]
```diff
@@ -73,7 +73,7 @@ mod tests {
...
```
Comment (niklaslong) [inline circuit/types/address/src/equal.rs]: And again.

File: circuit/types/field/src/double.rs [hunk 16]
```diff
@@ -51,9 +51,11 @@ mod tests {
     const ITERATIONS: u64 = 10_000;
 
     fn check_double(name: &str, mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/double.rs]: Ditto

File: circuit/types/field/src/helpers/to_bits.rs [hunk 17]
```diff
@@ -96,9 +96,11 @@ mod tests {
     fn check_to_bits_le(mode: Mode) {
         let expected_number_of_bits = console::Field::<<Circuit as Environment>::Network>::size_in_bits();
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/helpers/to_bits.rs:99]: ditto

File: circuit/types/field/src/helpers/to_bits.rs [hunk 18]
```diff
@@ -122,9 +124,11 @@ mod tests {
     fn check_to_bits_be(mode: Mode) {
         let expected_number_of_bits = console::Field::<<Circuit as Environment>::Network>::size_in_bits();
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/helpers/to_bits.rs:127]: ditto

File: circuit/types/field/src/inverse.rs [hunk 19]
```diff
@@ -78,9 +78,11 @@ mod tests {
     const ITERATIONS: u64 = 1_000;
 
     fn check_inverse(name: &str, mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/inverse.rs]: ditto

File: circuit/types/field/src/neg.rs [hunk 20]
```diff
@@ -75,9 +75,11 @@ mod tests {
             });
         };
 
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/neg.rs]: ditto

File: circuit/types/field/src/square.rs [hunk 21]
```diff
@@ -63,9 +63,11 @@ mod tests {
     }
 
     fn run_test(mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/square.rs]: ditto

File: circuit/types/field/src/square_root.rs [hunk 22]
```diff
@@ -62,9 +62,11 @@ mod tests {
     const ITERATIONS: u64 = 1_000;
 
     fn check_square_root(name: &str, mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/field/src/square_root.rs]: same here

File: circuit/types/integers/src/abs_checked.rs [hunk 23]
```diff
@@ -100,9 +100,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + UnwindSafe>(mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/abs_checked.rs:103]: Not entirely sure if or where this function is used but if it's called multiple times the rng should be passed in.

File: circuit/types/integers/src/abs_wrapped.rs [hunk 24]
```diff
@@ -90,9 +90,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + UnwindSafe>(mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/abs_wrapped.rs:93]: Same as above.

File: circuit/types/integers/src/add_checked.rs [hunk 25]
```diff
@@ -199,9 +199,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/add_checked.rs:202]: ditto

File: circuit/types/integers/src/add_wrapped.rs [hunk 26]
```diff
@@ -96,9 +96,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/add_wrapped.rs:99]: ditto

File: circuit/types/integers/src/and.rs [hunk 27]
```diff
@@ -140,9 +140,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + BitAnd<Output = I>>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/and.rs:143]: same here

File: circuit/types/integers/src/compare.rs [hunk 28]
```diff
@@ -159,9 +159,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/compare.rs:162]: yep, you guessed it 🙂

File: circuit/types/integers/src/div_checked.rs [hunk 29]
```diff
@@ -213,9 +213,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/div_checked.rs:216]: same here

File: circuit/types/integers/src/div_wrapped.rs [hunk 30]
```diff
@@ -200,9 +200,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/div_wrapped.rs:203]: ditto

File: circuit/types/integers/src/equal.rs [hunk 31]
```diff
@@ -102,9 +102,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/equal.rs:105]: ditto

File: circuit/types/integers/src/modulo.rs [hunk 32]
```diff
@@ -76,9 +76,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/modulo.rs:79]: ditto

File: circuit/types/integers/src/mul_checked.rs [hunk 33]
```diff
@@ -284,11 +284,13 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/mul_checked.rs:287]: same here

File: circuit/types/integers/src/mul_wrapped.rs [hunk 34]
```diff
@@ -113,11 +113,13 @@ mod tests {
     }
 
     fn run_test<I: IntegerType>(mode_a: Mode, mode_b: Mode) {
+        let rng = &mut TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/mul_wrapped.rs:116]: same

File: circuit/types/integers/src/neg.rs [hunk 35]
```diff
@@ -106,14 +106,16 @@ mod tests {
         // Check the 1 case.
         check_neg::<I>(&format!("Neg: {} one", mode), -console::Integer::one(), mode);
         // Check random values.
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/neg.rs:109]: ditto

File: circuit/types/integers/src/not.rs [hunk 36]
```diff
@@ -79,9 +79,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + Not<Output = I>>(mode: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/not.rs:82]: ditto

File: circuit/types/integers/src/or.rs [hunk 37]
```diff
@@ -138,9 +138,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + BitOr<Output = I>>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/or.rs:141]: same here

File: circuit/types/integers/src/pow_checked.rs [hunk 38]
```diff
@@ -194,9 +194,11 @@ mod tests {
     }
 
     fn run_test<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+        let mut rng = TestRng::default();
```
Comment (niklaslong) [inline circuit/types/integers/src/pow_checked.rs:197]: same here

File: circuit/types/field/src/helpers/to_bits.rs (hunk 17 above)
Comment (ljedrz) [reply circuit/types/field/src/helpers/to_bits.rs:99]: this one was actually fine, as the consecutive calls were for `check_bits_le`

File: circuit/types/field/src/helpers/to_bits.rs (hunk 18 above)
Comment (ljedrz) [reply circuit/types/field/src/helpers/to_bits.rs:127]: this one was actually fine, as the consecutive calls were for `check_bits_be`

File: circuit/types/integers/src/abs_checked.rs (hunk 23 above)
Comment (ljedrz) [reply circuit/types/integers/src/abs_checked.rs:103]: this one's fine, as it creates an individual test case every time it's called

**Comment by ljedrz** [comment]
The CI error seems unrelated:
```
ATTENTION - "universal.srs.trial.3747e59" does not exist, downloading this file remotely and storing it locally. Please ensure "universal.srs.trial.3747e59" is stored in "/home/circleci/.aleo/resources/universal.srs.trial.3747e59".

snarkvm_parameters::testnet3 - Downloading parameters...
snarkvm_parameters::testnet3 - 2.72% complete (294 MB total)thread 'ledger::block::transactions::serialize::tests::test_bincode' panicked at 'Failed to load universal SRS bytes: Crate("curl::error", "Error { description: \"Transferred a partial file\", code: 18, extra: Some(\"transfer closed with 299937167 bytes remaining to read\") }")', vm/compiler/src/snark/universal_srs.rs:77:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    ledger::block::transactions::serialize::tests::test_bincode

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 669 filtered out; finished in 76.15s

error: test failed, to rerun pass '--lib'

Exited with code exit status 101
```

Will retry once other jobs are complete.

**Comment by ljedrz** [comment]
The errors are network-related again; I'll retry the affected jobs at a later time.

**Comment by ljedrz** [comment]
Open point: there are some fixed Rngs used in several spots, and while this PR preserves them, they should probably not be fixed for better coverage.

**Review by howardwu — APPROVED** [review]
LGTM, that took a while to parse, but it looks great! Thanks for the work.

To my knowledge, the only breaking change is the universal SRS for phase 1 was previously generated from `StdRng::seed_from_u64(1245897092u64)` and is now generated from `XorShiftRng::seed_from_u64(1245897092u64)`, which is fine.

**Pull request merged**